src/SArray.jl

```diff
@@ -201,12 +201,43 @@ function static_array_gen(::Type{SA}, @nospecialize(ex), mod::Module) where {SA}
             if length(ex.args) == 1
                 f === :zeros || f === :ones || error("@$SA got bad expression: $(ex)")
                 return :($f($SA{$Tuple{},$Float64}))
-            end
-            return quote
-                if isa($(esc(ex.args[2])), DataType)
-                    $f($SA{$Tuple{$(escall(ex.args[3:end])...)},$(esc(ex.args[2]))})
-                else
-                    $f($SA{$Tuple{$(escall(ex.args[2:end])...)}})
+            elseif f !== :rand || length(ex.args) == 2
+                return quote
+                    if isa($(esc(ex.args[2])), DataType)
```

Comment:
Rather than dispatching with an `if` statement, it seems a lot easier to simply rewrite to a custom function call and then using ordinary dispatch.   See my comment https://github.com/JuliaArrays/StaticArrays.jl/issues/804#issuecomment-1834474456

I used a similar strategy in ChangePrecision.jl — https://github.com/JuliaMath/ChangePrecision.jl/blob/1efe76bfa9e4bf47f9bfe247bc6fe5568ddad64b/src/ChangePrecision.jl#L20-L24 — and it's very pleasant and flexible in my experience.

Reply:
Thanks, that's a very good suggestion.

Reply:
@hyrodium implemented it in #1213